pkg/export/gcm/promtest/prometheus.go

```diff
@@ -0,0 +1,179 @@
+package promtest
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/efficientgo/e2e"
+	e2emon "github.com/efficientgo/e2e/monitoring"
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type promBackend struct {
+	image string
+
+	g *recordGatherer
+	p *e2emon.Prometheus
+}
+
+func Prometheus(image string) Backend {
+	return &promBackend{image: image}
+}
+
+func (p *promBackend) Ref() string { return "prometheus" }
+
+// recordGatherer is a prometheus.Gatherer capable to "play" the recorded metric state
+// with fixed timestamp to the backfill data into Prometheus compatible system.
+type recordGatherer struct {
+	i              int
+	plannedScrapes [][]*dto.MetricFamily
+	mu             sync.Mutex
+}
+
+func (g *recordGatherer) Gather() (ret []*dto.MetricFamily, _ error) {
+	g.mu.Lock()
+	ret = nil
+	if g.i > -1 && g.i < len(g.plannedScrapes) {
+		ret = g.plannedScrapes[g.i]
+		g.i++
+	}
+	g.mu.Unlock()
+
+	return ret, nil
+}
+
+func (p *promBackend) start(t testing.TB, env e2e.Environment) (v1.API, map[string]string) {
+	t.Helper()
+
+	const name = "prometheus"
+
+	m := http.NewServeMux()
+	p.g = &recordGatherer{i: -1}
+	m.Handle("/metrics", promhttp.HandlerFor(p.g, promhttp.HandlerOpts{
+		EnableOpenMetrics: true,
+	}))
+
+	// Listen on all addresses, since we need to connect to it from docker container.
+	list, err := net.Listen("tcp", "0.0.0.0:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, port, err := net.SplitHostPort(list.Addr().String())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	s := http.Server{Handler: m}
+
+	go func() { _ = s.Serve(list) }()
+	env.AddCloser(func() { _ = s.Close() })
+
+	p.p = newPrometheus(env, name, p.image, net.JoinHostPort(env.HostAddr(), port), nil)
+	if err := e2e.StartAndWaitReady(p.p); err != nil {
+		t.Fatalf("can't start Prometheus: %v", err)
+	}
+
+	cl, err := api.NewClient(api.Config{Address: "http://" + p.p.Endpoint("http")})
+	if err != nil {
+		t.Fatalf("create Prometheus client: %s", err)
+	}
+
+	return v1.NewAPI(cl), map[string]string{"job": "test"}
+}
+
+func newPrometheus(env e2e.Environment, name string, image string, scrapeTargetAddress string, flagOverride map[string]string) *e2emon.Prometheus {
+
+	if image == "" {
+		image = "quay.io/prometheus/prometheus:v2.47.2"
```

Comment:
Would it maybe make sense to make this a flag or env var, that way we can run the test against different Prometheus versions without changing code? Might be useful for quick comparisons.

Reply:
Nice! Yea we can do that, but perhaps we could do that in the PR/event where somebody actually would like to do such test?

Reply:
SGTM!